service/src/main/java/com/commercetools/pspadapter/payone/transaction/PaymentMethodDispatcher.java

```diff
@@ -1,37 +1,41 @@
 package com.commercetools.pspadapter.payone.transaction;
 
+import com.commercetools.payments.TransactionStateResolver;
 import com.commercetools.pspadapter.payone.domain.ctp.PaymentWithCartLike;
 import com.google.common.collect.ImmutableMap;
 import io.sphere.sdk.payments.Transaction;
-import io.sphere.sdk.payments.TransactionState;
 import io.sphere.sdk.payments.TransactionType;
 
+import javax.annotation.Nonnull;
+
 public class PaymentMethodDispatcher {
     private final TransactionExecutor defaultExecutor;
     private final ImmutableMap<TransactionType, TransactionExecutor> executors;
+    private final TransactionStateResolver transactionStateResolver;
 
     public PaymentMethodDispatcher(
-            final TransactionExecutor defaultExecutor,
-            final ImmutableMap<TransactionType, TransactionExecutor> executors) {
+            @Nonnull final TransactionExecutor defaultExecutor,
+            @Nonnull final ImmutableMap<TransactionType, TransactionExecutor> executors,
+            @Nonnull final TransactionStateResolver transactionStateResolver) {
         this.defaultExecutor = defaultExecutor;
         this.executors = executors;
+        this.transactionStateResolver = transactionStateResolver;
     }
 
-    public PaymentWithCartLike dispatchPayment(final PaymentWithCartLike paymentWithCartLike) {
-        // Execute the first Pending Transaction
+    public PaymentWithCartLike dispatchPayment(@Nonnull final PaymentWithCartLike paymentWithCartLike) {
+        // Execute the first uncompleted transaction
         return paymentWithCartLike.getPayment()
             .getTransactions()
             .stream()
-            .filter(transaction -> TransactionState.PENDING.equals(transaction.getState()))
+            .filter(transactionStateResolver::isNotCompletedTransaction)
             .findFirst()
             .map(transaction -> {
                 final PaymentWithCartLike newPaymentWithCartLike = executeTransaction(paymentWithCartLike, transaction);
                 final Transaction updatedTransaction = getUpdatedTransaction(transaction, newPaymentWithCartLike);
-
-                if (TransactionState.PENDING.equals(updatedTransaction.getState())) { // Still Pending, stop ->
-                    // executor has done its duty, notification from Payone required to proceed with this transaction
+                if (transactionStateResolver.isNotCompletedTransaction(updatedTransaction)) { // Still not completed ->
+                    // stop, executor has done its duty, notification from Payone required to proceed with this transaction
                     return newPaymentWithCartLike;
```

Comment:
1) we support now only authorization and charge transactions, no refund
2) i'm not sure how the service should work in case of multiple transactions. should we process them all? then maybe this implementation was wrong from the beginning

Reply:
What about this case: T1: Authorisation, T2: Charge? 

I do not have valid business use case for scenarios with multiple unprocessed transactions but it does not seem right that we neither validate/prohibit multiple unprocessed transactions (something like IllegalTransactionAmountException) nor we ensure that we process all of them. As long as we do not have valid use case i would vote for validation/exception.

Reply:
added as re-factoring point:
https://github.com/commercetools/commercetools-payone-integration/issues/208